### PR TITLE
Top collector

### LINKED ISSUE #388: Rename the current `TopCollector` into `TopScoreCollector`, and introduce a `TopFieldCollector`
Currently there is no off the shelf solution to sort by a given field. (e.g. date, price), instead of the BM25 relevancy score.

Following Lucene, the current `TopCollector` collector could be renamed `TopScoreCollector`, and the new collector could be called `TopFieldCollector`. If possible the two somehow share some code in common...

### PR DESCRIPTION
**Pull request by pentlander**
Closes #388

### CODE REVIEW

File: src/collector/top_collector.rs [hunk 1]
```diff
@@ -150,9 +158,56 @@ impl TopCollector {
...
+    pub fn score_docs(&self) -> Vec<(Score, DocAddress)> {
+        self.feature_docs()
+    }
 }
 
-impl Collector for TopCollector {
+// This will be removed in a future commit, as there cannot be an impl for just score as well as
+// for T of type PartialOrd. This would be possible if specialization were stable.
+#[deprecated(since = "0.1")]
+impl Collector for TopCollector<Score> {
```
Comment (fulmicoton) [inline src/collector/top_collector.rs]: It's ok to deprecate it entirely as long as we have not reached 1.0

**Comment by fulmicoton** [comment]
Overall that looks good! There is a bit of PR stack. Let me come back to you later this week.

File: src/collector/mod.rs [hunk 2]
```diff
@@ -17,6 +17,12 @@ pub use self::multi_collector::MultiCollector;
 mod top_collector;
 pub use self::top_collector::TopCollector;
```
Comment (fulmicoton) [inline src/collector/mod.rs]: Let's keep `TopCollector` crate private. 

In exchange, for backward probabilitiy we can

```rust
#[deprecated]
pub use TopScoreCollector as TopCollector;
```

Comment (fulmicoton) [inline src/collector/mod.rs]: Let's keep `TopCollector` crate private. 

In exchange, for backward probabilitiy we can

```rust
#[deprecated]
pub use self::top_score_collector::TopScoreCollector as TopCollector;
```

File: src/collector/multi_collector.rs [hunk 3]
```diff
@@ -104,7 +104,7 @@ mod tests {
 
     #[test]
     fn test_multi_collector() {
-        let mut top_collector = TopCollector::with_limit(2);
+        let mut top_collector: TopCollector<Score> = TopCollector::with_limit(2);
```
Comment (fulmicoton) [inline src/collector/multi_collector.rs]: Use `TopScoreCollector`

File: src/collector/top_collector.rs [hunk 4]
```diff
@@ -150,9 +158,56 @@ impl TopCollector {
...
+    /// Sets the segment local ID for the collector
+    pub fn set_segment_id(&mut self, segment_id: SegmentLocalId) {
+        self.segment_id = segment_id;
+    }
+
+    /// Collects a document scored by the given feature
+    ///
+    /// It collects documents until it has reached the max capacity. Once it reaches capacity, it
+    /// will compare the lowest scoring item with the given one and keep whichever is greater.
+    pub fn collect_feature(&mut self, doc: DocId, feature: T) {
```
Comment (fulmicoton) [inline src/collector/top_collector.rs]: does not need to be `pub`

File: src/collector/top_field_collector.rs [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+///             rating => 80u64,
+///         ));
+///         index_writer.commit().unwrap();
+///     }
+///
+///     index.load_searchers()?;
+///     let searcher = index.searcher();
+///
+///     {
+///	        let mut top_collector = TopFieldCollector::with_limit("rating", 2);
```
Comment (fulmicoton) [inline src/collector/top_field_collector.rs]: indent

File: src/collector/top_field_collector.rs [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+impl<T: FastValue + PartialOrd + Clone> TopFieldCollector<T> {
+    /// Creates a top field collector, with a number of documents equal to "limit".
+    ///
+    /// The given field name must be a fast field, otherwise the collector have an error while
+    /// collecting results.
+    ///
+    /// # Panics
+    /// The method panics if limit is 0
+    pub fn with_limit(field_name: impl Into<String>, limit: usize) -> Self {
```
Comment (fulmicoton) [inline src/collector/top_field_collector.rs]: For consistency, use `field: Field`. 

I agree this kind of sucks. I think at one point, we'll have to make `Field` `String`-based and use `Into<Field>`.

File: src/collector/top_field_collector.rs [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+    /// The result of the sort is not cached.
+    pub fn docs(&self) -> Vec<DocAddress> {
+        self.collector.docs()
+    }
+
+    /// Returns K best FieldDocuments sorted in decreasing order.
+    ///
+    /// Calling this method triggers the sort.
+    /// The result of the sort is not cached.
+    pub fn field_docs(&self) -> Vec<(T, DocAddress)> {
```
Comment (fulmicoton) [inline src/collector/top_field_collector.rs]: Rename `top_docs()`

File: src/collector/top_collector.rs [hunk 8]
```diff
@@ -125,22 +128,27 @@ impl TopCollector {
...
-    /// Returns K best ScoredDocument sorted in decreasing order.
+    /// Returns K best FeatureDocuments sorted in decreasing order.
     ///
     /// Calling this method triggers the sort.
     /// The result of the sort is not cached.
-    pub fn score_docs(&self) -> Vec<(Score, DocAddress)> {
-        let mut scored_docs: Vec<GlobalScoredDoc> = self.heap.iter().cloned().collect();
-        scored_docs.sort();
-        scored_docs
+    pub fn feature_docs(&self) -> Vec<(T, DocAddress)> {
```
Comment (fulmicoton) [inline src/collector/top_collector.rs]: Rename top_docs()

File: src/collector/top_score_collector.rs [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    /// The result of the sort is not cached.
+    pub fn docs(&self) -> Vec<DocAddress> {
+        self.collector.docs()
+    }
+
+    /// Returns K best ScoredDocuments sorted in decreasing order.
+    ///
+    /// Calling this method triggers the sort.
+    /// The result of the sort is not cached.
+    pub fn score_docs(&self) -> Vec<(Score, DocAddress)> {
```
Comment (fulmicoton) [inline src/collector/top_score_collector.rs]: rename `top_docs()`. Keep `score_docs` and mark as deprecated.

**Review by fulmicoton — CHANGES_REQUESTED** [review]
See comments in the CR.

File: src/collector/top_field_collector.rs (hunk 5 above)
Comment (pentlander) [reply src/collector/top_field_collector.rs]: Strange, it isn't like this in my workspace.

File: src/collector/mod.rs [hunk 10]
```diff
@@ -15,10 +15,11 @@ mod multi_collector;
 pub use self::multi_collector::MultiCollector;
 
 mod top_collector;
-pub use self::top_collector::TopCollector;
 
 mod top_score_collector;
 pub use self::top_score_collector::TopScoreCollector;
+#[deprecated]
+pub use self::top_score_collector::TopScoreCollector as TopCollector;
```
Comment (fulmicoton) [inline src/collector/mod.rs]: thx

**Comment by pentlander** [comment]
The build is failing but it looks like the master is already failing?

**Comment by fulmicoton** [comment]
@pentlander You are right my bad. Fixing it.

**Pull request merged**